### ISSUE TITLE
pkzip salt problem

### ISSUE THREAD

**Issue by magnumripper** [magnumripper/JohnTheRipper#1862]
In the [fakedb_core](https://github.com/magnumripper/JohnTheRipper/commits/fakedb_core) topic branch, we get this:
```
$ ../run/john -test=0 -form:pkzip -verb:4
Testing: PKZIP [32/64]... (8xOMP) Loaded 9 hashes with 9 different salts to test db from test vectors
FAILED (Could not find salt in db - salt() return inconsistent?)
```
This is the only format failing. What happens is we load a "fake db" from the format's test vectors, then during self-test we look up the db->salts->salt that matches current test vector's salt. And in this one case, we don't find it.

Dynamic salts are honored (and works fine with other formats) yet pkzip somehow fails. I have no idea what is wrong but I suspect it's a bug that doesn't surface in the main branch but that will likely bust pot sync and perhaps other things. @jfoug?

**Comment by magnumripper** [comment]
BTW the same code also found a bunch of other salt cleaning problems, fixed in 7f0fd4f7e. I'm not sure why our current tests did not catch them.

**Comment by magnumripper** [comment]
@jfoug please review d41cc16. I think I got it right: The problem was that set_salt() does some late work, modifying the salt database. Not sure why it does that, but I moved that struct member out of what is included in `dyna_salt_cmp()`.
